### PR TITLE
More juice httpd collector

### PR DESCRIPTION
**Pull request by tsheasha**
Reporting the Total Apache Acceses per Second as a metric, as well as adding stats on CPULoad of the apache process and the memory usage of the workers *resident set size and virtual memory size".

### CODE REVIEW

File: src/diamond/collectors/httpd/httpd.py [hunk 1]
```diff
@@ -117,53 +126,111 @@ def collect(self):
...
+        except Exception as e:
+            self.log.error(
+                "Failed because: {0!s}".format(e)
+            )
+
     def _publish(self, nickname, key, value):
 
-        metrics = ['ReqPerSec', 'BytesPerSec', 'BytesPerReq', 'BusyWorkers',
+        metrics = ['AccessesPerSec', 'ReqPerSec', 'BytesPerSec', 'BytesPerReq', 'BusyWorkers',
                    'Total Accesses', 'IdleWorkers', 'StartingWorkers',
```
Comment (poros) [inline src/diamond/collectors/httpd/httpd.py]: Could we avoid "Total Accesses" in favor of "TotalAccesses"?

File: src/diamond/collectors/httpd/httpd.py [hunk 2]
```diff
@@ -117,53 +126,111 @@ def collect(self):
                         else:
                             self._publish(nickname, k, v)
 
+                        if k == 'Total Accesses':
+                            gen_metric_name = 'AccessesPerSec'
+                            gen_metric_value = 0
+                            if self.last_collected_time and self.last_collected_total_accesses:
```
Comment (poros) [inline src/diamond/collectors/httpd/httpd.py]: nit: is not None

File: src/diamond/collectors/httpd/httpd.py [hunk 3]
```diff
@@ -117,53 +126,111 @@ def collect(self):
...
+                )
+            else:
+                resident_memory = {}
+                virtual_memory = {}
+                for line in output.split('\n'):
+                    if not line:
+                        continue
+                    (rss, vsz, proc) = line.strip('\n').split(None,2)
+                    if proc in self.config['processes']:
+                        resident_memory.setdefault(proc, []).append(int(rss))
```
Comment (poros) [inline src/diamond/collectors/httpd/httpd.py]: why not collections.defaultdict?

**Comment by poros** [comment]
Unfortunately the code of the httpd collector is poorly structured, but it's inherited from diamond and I don't see this review as the right place to fix it (even though it adds a lot more complexity to the collector).

The direct consequence of this fact is that adding tests for these new functionalities would be very painful :(

Having said that, I am in favor or merging this PR
+1

**Comment by tsheasha** [comment]
@poros I don't think changing the structure of the collectors, even we think they are poorly structured, is a good idea. This is due to that fact that we want to upstream the least changes to the core diamond core architecture as much as possible, we just append new functionalities to it. That is the way I see it at leat, which could be complete non-sense.
Let me know what you think about that.

As for the complexity, I don't really get what you referring to, can you be more concrete on that?

File: src/diamond/collectors/httpd/httpd.py (hunk 3 above)
Comment (tsheasha) [inline src/diamond/collectors/httpd/httpd.py]: I was not aware of that function, I will use it instead. Thanks

**Comment by poros** [comment]
With complexity I mean that you are adding functionality to this collector (you're adding a metric which requires state and you're spawning a subprocess). It's all legit, but it would be great if we were able to test them easily. We are not, sadly.

About the fixing the poor structure, I agree with you, that's why I +1ed the PR. I am unsure about the plans to contribute to/pull from new versions of diamond. It's something that has never been addressed until now.

**Comment by poros** [comment]
+1

**Comment by zachm** [comment]
Signed off. Merging!

**Pull request merged**